benches/backtracking.rs

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This bench monitors the performance of backtracking and term intersection.
+//!
+//! Dependencies are constructed in a way that all versions need to be tested before
+//! pubgrub can determine that no solution exists.
+
+use criterion::*;
+use pubgrub::OfflineDependencyProvider;
+use version_ranges::Ranges;
+
+fn bench(c: &mut Criterion, case: &str, package_count: u32, version_count: u32) {
+    let mut dependency_provider = OfflineDependencyProvider::<u32, Ranges<u32>>::new();
+    let r = Ranges::strictly_lower_than(version_count);
+    dependency_provider.add_dependencies(0u32, 0u32, [(1u32, r)]);
+
+    for n in 1..package_count {
+        for v in 0..version_count {
+            let r = Ranges::strictly_lower_than(v);
```

Comment:
This  pattern of layers of crates that each depend on the next can easily cause exponential  resolution time in naïve backtracking resolvers. Some variation on this is definitely a benchmark we should keep an eye on.

Why is this `strictly_lower_than` instead of `higher_than`?

I would imagine that  if were using `strictly_lower_than` and  we are processing version 9,  then we have to check  the dependency `(n + 1, <9)`,  which we discover doesn't work.  then when we go on to processing version 8 we now have to process `(n + 1, <8)`. This only takes a  constant amount of time  because `<8`  is a subset of `<9`.  On the other hand  if we used `higher_than`  and we process version 9,  then we have to check  the dependency `(n + 1, >=9)`, which we discover doesn't work.  then when we go on to processing version 8 we now have to process `(n + 1, >=8)`. Which involves actually having to go do some work to figure out there's a problem with the 8 we have not looked at befor.

If `strictly_lower_than`  is slower than `higher_than`  I would very much want to know why.

That same question could also be asked as  how is this like the realistic  Solana example?  For most of those crates we have the pattern `n, v, [(n + 1, ^v)` Where `^`  basically de-sugars to `>=`. Could you explain why  this synthetic benchmark will have similar performance characteristics to Solana?

Reply:
This benchmark reproduces the backtracking behavior of solana and is close to the test that I recently added in cargo: https://github.com/Eh2406/pubgrub-crates-benchmark/issues/6#issuecomment-2400296431

Reply:
Interestingly the  flame graph provided by codspeed https://codspeed.io/pubgrub-rs/pubgrub/branches/x-hgg-x:new-bench shows this benchmark spending most of its time in `satisfier_search`, which  if I remember correctly is not where `Solana` spends its time.

Reply:
I committed new benchmarks representing various other patterns found in solana.